did_core/did_parser_nom/tests/did/negative.rs

```diff
@@ -0,0 +1,27 @@
+use did_parser_nom::Did;
+
+macro_rules! test_cases_negative {
+    ($($name:ident: $input:expr)*) => {
+        $(
+            #[test]
+            fn $name() {
+                println!("Testing {}", $input);
+                assert!(Did::parse($input.to_string()).is_err());
+            }
+        )*
+    };
+}
+
+test_cases_negative! {
+    test_failure_case1: ""
+    test_failure_case2: "not-a-did"
+    test_failure_case3: "did:example"
+    test_failure_case4: "2ZHFFhtTD6hJqzux"
+    test_failure_case5: "did:indy:s@vrin:7Tqg6BwSSWapxgUDm9KKgg"
+    test_failure_case6: "did:indy:sovrin:alpha:%0zqg6BwS.Wapxg-Dm9K_gg"
+    test_failure_case7: "did:sov:2wJPyULfLLnYTEFYzByf"
+    test_failure_case8: "did:sov:2wJPyULfOLnYTEFYzByfUR"
+    test_failure_case9: "2wJPyULfLLnYTEFYzByf"
+    test_failure_case10: "2wJPyULfOLnYTEFYzByfUR"
```

Comment:
what's the difference between these two, given they both are negative?

Reply:
What do you mean by difference? :) They are strings, isn't the difference obvious? :)

Reply:
Both input fail, but the inputs are slightly difference. What am I missing?

Reply:
No problem, let's do it

Reply:
This discussion got duplicated, I am here quoting from thread https://github.com/hyperledger/aries-vcx/pull/1099#discussion_r1457413330

> To understand what makes a DID invalid you should understand what makes it valid, which at the very least means looking at the positive test cases, preferably the method's specification; there is no way to explain the same in a test name.

> At best, the test name can give an illusion of understanding or partial explanation (like test_sov_unqualified_invalid_len or test_sov_unqualified_invalid_char), at the cost of long, hard-to-read and potentially confusing test names (like test_indy_non_method_specific_id_char_in_namespace).

> I understand and am sympathetic to the argument that (or test comments) is better than no explanation / documentation at all.

I think the names you suggested are better then simply `negative_test_x`. Yes, the example `test_indy_non_method_specific_id_char_in_namespace` is taken to the other extreme of the spectrum. I don't think it has to be that explicit, but at least brief indication what the test is testing can be helpful to reader.

Reply:
I gave names to the negative cases, let me know if they are good enough

Reply:
looks good! Thanks